### PR TITLE
Add strict typing part 1

### PR DESCRIPTION
**Pull request by janiversen**
The files are not selected by random, I have made a dependency tree, and the files in this PR depend on nothing or on each other and therefore can be converted without needing to change other files.

The most important file is command.py, with a lot of changes that affect the bulk of the library as well as external use, this file need an extra careful review please.

In order to control that command.py is typed correctly, I have converted around 50% of the library and the findings flowed back into command.py. I have marked a number of code changes for a later PR, since I do not want to mix "themes".

Once this PR is corrected and merged, I will adapt the changes in the next batch and submit it. A batch is defined as:
- files that only depend on each other or already converted files
- the number of changes are below 200
- preferable only one file with complicated conversion

I can easily adapt the batch definition and submit depending on your preference.

### CODE REVIEW

**Comment by janiversen** [comment]
@MartinHjelmare all yours :smile:

File: pytradfri/command.py [hunk 1]
```diff
@@ -83,12 +87,16 @@ def path_str(self) -> str:
...
         """Generate url for coap client."""
         return "coaps://{}:5684/{}".format(host, self.path_str)
 
-    def _merge(self, a, b):
+    def _merge(self, a: Optional[dict[str, Any]], b: Optional[dict[str, Any]]) -> Optional[dict[str, Any]]:
         """Merges a into b."""
+        if a is None:
+            return b
+        if b is None:
+            return b
```
Comment (KapJI) [inline pytradfri/command.py:99]:
```suggestion
            return a
```

**Comment by janiversen** [comment]
Hmmm seems the githooks needs to be modified, I am so used to having the githook do formatting etc. that I did not think about doing it.

File: pytradfri/command.py (hunk 1 above)
Comment (janiversen) [reply pytradfri/command.py]: Actually that causes a test to fail (I started by doing that). But anyhow I have changed it to be easier to understand.

Comment (janiversen) [reply pytradfri/command.py]: You are right, but when you look at the callers the parameters are optional and can be None (even though only in error situations), This is one of the cases, where adding typing helped solve a potential bug.

Comment (janiversen) [reply pytradfri/command.py]: I am trying to touch the code as little as possible while adding the typing (and making a note, where to "enhance" the code later). The code is generally written for "good weather", there are a number of places like this (and I have seen enough issues in cases where it was "bad weather").

**Comment by janiversen** [comment]
@KapJI if/when you have time I would really like your input on how solve this:
https://github.com/home-assistant-libs/pytradfri/blob/master/pytradfri/smart_task.py line 37 (definition)
mypy complains:
```
pytradfri/smart_task.py:38: error: Argument 1 to "BitChoices" has incompatible type "Tuple[Tuple[str, str], Tuple[str, str], Tuple[str, str], Tuple[str, str], Tuple[str, str], Tuple[str, str], Tuple[str, str]]"; expected "Tuple[Tuple[str, str]]"  [arg-type]
```
I have defined BitChoise (in util.py) as:
```
class BitChoices(object):
    """Helper class for bitwise dates.

    http://stackoverflow.com/questions/3663898/representing-a-multi-select-field-for-weekdays-in-a-django-model
    """

    def __init__(self, choices: tuple[tuple[str,str]]) -> None:
```
I think my definition is correct, but it WEEKDAY is defined with each entry in the tuple.

Any idea/advice is mostly welcome. Thanks in advance.

**Comment by KapJI** [comment]
Use `...`: `choices: Tuple[Tuple[str,str], ...]`

**Comment by janiversen** [comment]
@KapJI do you have more comments or can you approve the PR ?

File: pytradfri/command.py [hunk 2]
```diff
@@ -1,23 +1,28 @@
...
     """The object for coap commands."""
 
     def __init__(
         self,
-        method,
-        path,
-        data=None,
+        method: str,
+        path: list[str],
+        data: Optional[dict[str, Any]] = None,
```
Comment (KapJI) [inline pytradfri/command.py:18]: Don't mix old and new style annotations.
```suggestion
        data: dict[str, Any] | None = None,
```

File: pytradfri/command.py [hunk 3]
```diff
@@ -26,52 +31,52 @@ def __init__(
         self._err_callback = err_callback
         self._observe = observe
         self._observe_duration = observe_duration
-        self._raw_result = None
-        self._result = None
+        self._raw_result: Optional[str] = None
+        self._result: Optional[str] = None
```
Comment (KapJI) [inline pytradfri/command.py:35]:
```suggestion
        self._raw_result: str | None = None
        self._result: str | None = None
```

File: pytradfri/command.py [hunk 4]
```diff
@@ -26,52 +31,52 @@ def __init__(
...
         return self._method
 
     @property
-    def path(self):
+    def path(self) -> list[str]:
         return self._path
 
     @property
-    def data(self):
+    def data(self) -> Optional[dict[str, Any]]:
```
Comment (KapJI) [inline pytradfri/command.py:46]: You get the idea, there're few more places.
```suggestion
    def data(self) -> dict[str, Any] | None:
```

File: pytradfri/command.py [hunk 5]
```diff
@@ -1,23 +1,28 @@
 """Command implementation."""
+from __future__ import annotations
 
 from copy import deepcopy
+from typing import Any, Optional, Callable
+
+TYPE_PROCESS_RESULT_CB = Optional[Callable[[Any], Optional[str]]]
+TYPE_ERR_CB = Optional[Callable[[str], str]]
```
Comment (KapJI) [inline pytradfri/command.py:8]: The old style annotations should be only used here. Until Python 3.10 is a default.

File: pytradfri/command.py (hunk 2 above)
Comment (janiversen) [reply pytradfri/command.py]: I did not see it as old/new, but you are right they are all changed.

File: pytradfri/command.py (hunk 5 above)
Comment (janiversen) [reply pytradfri/command.py:8]: Yup these do not work (yet) with the new style.

**Review by KapJI — APPROVED** [review]
LGTM

**Comment by janiversen** [comment]
Thanks.

**Comment by janiversen** [comment]
@MartinHjelmare do you have comments?

**Review by MartinHjelmare — APPROVED** [review]
Nice!

**Pull request merged**